wikimedia/wikipedia-ios
# Make WMFContentGroup and WMFArticle language variant-aware

This is work towards the Chinese variants feature https://phabricator.wikimedia.org/T195265
It is not the entire feature/ticket.

Each commit is self-contained it is probably easiest to review them individually.

### Notes
**Commit 1**
In WMFContentGroup URL methods, propagates the language variant to the created URL.

In cases where an argument needed to be added, only the language variant is passed in, not a source URL. This is to indicate that those methods do not require a source URL, only the language variant.

Call sites are updated where necessary.

**Commit 2**
Update appropriate WMFContentGroup fetch methods to include the language variant as part of the predicate.

**Commit 3**
Clean up how the variant value is assigned in WMFContentGroup. Ensure URL, siteURL, and articleURL properties propagate the variant. Assert that any set URL value has the same variant value as the content group.

**Commit 4**
Add a class to serve as a compound key in the NSCache article cache in the MWKDataStore. Add a test to ensure it works as expected as a key.

**Commit 5**
Update WMFArticle and MWKDataStore to include the variant in fetches and use the new compound key for NSCache operations.

Note a few non-language variant-aware methods are left as temporary shims. Once those subsystems are variant-aware, the shim methods will be removed.

**Commit 6**
Refactor of opportunity. The deleted MWKDataStore methods are unused.

**Commit 7**
The 'contentPreview' property of WMFContentGroup and 'object' property of WMFContent are archived NSURLs, WMFMTLModel subclasses, or collections of those types. Since archived NSURLs have no notion of a language variant, the variant must be propagated to the unarchived objects. This is done in -awakeFromFetch so it only occurs for fetched objects once after fetch.

**Commit 8**
This commit makes sure Random articles do not appear as blank in the explore feed if you turn on language variants. It is a temporary shim until ArticleSummary is made variant-aware. This only has an effect if language variants are turned on.

### Test Steps
1. With variants turned off (current default) there should be no changes in app behavior.
2. With variants turned on, if you set multiple variants for the same language, you should see a set of explore feed cards for each variant. Content of cards should be visible on first launch and after quit and restart.
NOTE: The contents of the cards will not yet reflect the proper variant.
